exquisite_corpus/spm_naive_language_model.py

```diff
@@ -0,0 +1,428 @@
+            hidden_states.view(full_num_layers, batch_size, self.rnn.hidden_size),
+        )
+        output = self.decoder(output)
+        output = self.log_softmax(output)
+        if return_hidden:
+            new_hidden_states = new_hidden_states.view(hidden_states.size())
+            return output, new_hidden_states
+        else:
+            return output
+
+
+class LossFunction(nn.Module):
+    """
+    Module to compute the aggregate (over all elements of a batch, and all
+    sequence positions within each element) loss.
+    """
+
+    def __init__(self, reduction="mean"):
+        super().__init__()
+        self.loss_fn = nn.NLLLoss(reduction=reduction)
+
+    def forward(self, input, target):
+        return self.loss_fn(input.view(target.numel(), -1), target.view(-1))
+
+
+class ModelManager:
+    """
+    Objects that own torch models (modules) and provide convenience methods
+    for their training and use.
+    """
+
+    def __init__(
+        self,
+        model=None,
+        device=torch.device("cuda:0"),
+        **kwargs
+    ):
+        if model is None:
+            self.model = LanguageModel(**kwargs)
+        else:
+            self.model = model
+        self.device = device
+
+    @property
+    def device(self):
+        return self._device
+
+    @device.setter
+    def device(self, new_device):
+        if isinstance(new_device, str):
+            new_device = torch.device(new_device)
+        if new_device != torch.device("cpu") and not torch.cuda.is_available():
+            print("Warning:  cuda not available so setting device to CPU.")
+            self._device = torch.device("cpu")
+        else:
+            self._device = new_device
+        self.model.to(self._device)
+
+    def collate_batch_with_labels(self, raw_batch):
+        """
+        Take a raw batch (as returned by a DataLoader and SpmIdsBatchSampler),
+        collate it, and split off labels.
+        """
+        data = torch.LongTensor(np.vstack([x[:-1] for x in raw_batch]))
+        labels = torch.LongTensor(np.vstack([x[1:] for x in raw_batch]))
+        return data, labels
+
+    def train(
+        self,
+        train_dataset_path,
+        validation_dataset_path=None,
+        save_path=None,
+        min_length=2,
+        n_epochs=np.inf,
+        n_batches_between_saves=1000,
+        n_batches_between_messages=50,
+        n_batches_between_validations=25000,
+        batch_size=256,
+        n_data_loader_workers=0,
+        lr=0.05,
+        device=None,
+    ):
+        if device is not None:
+            self.device = device
+
+        self.model.train()
+        if min_length < 2:
+            raise ValueError("Cannot train with sentences shorter than 2 tokens.")
+        train_data = SpmIdsDataset(train_dataset_path, min_length=min_length)
+        sampler = SpmIdsBatchSampler(train_data, batch_size=batch_size)
+        pin_memory = self.device != torch.device("cpu")
+        data_loader = DataLoader(
```

Comment:
I think this invocation of the DataLoader must be what I was looking for. However, it's buried in the training method, and then repeated again in testing and such. Shouldn't this be its own method that can be used externally?

Reply:
Yeah, good call; the batch makers I mentioned above wrap all this logic up in one place.